routers/api/actions/artifacts.go

```diff
@@ -0,0 +1,473 @@
+	prefix string
+	fs     storage.ObjectStorage
+}
+
+func (ar artifactRoutes) buildArtifactURL(taskID, artifactID int64, suffix string) (string, error) {
+	uploadURL := strings.TrimSuffix(setting.AppURL, "/") + strings.TrimSuffix(ar.prefix, "/") +
+		strings.ReplaceAll(artifactRouteBase, "{taskID}", strconv.FormatInt(taskID, 10)) +
+		"/" + strconv.FormatInt(artifactID, 10) + "/" + suffix
+	u, err := url.Parse(uploadURL)
+	if err != nil {
+		return "", err
+	}
+	return u.String(), nil
+}
+
+// getUploadArtifactURL generates a URL for uploading an artifact
+func (ar artifactRoutes) getUploadArtifactURL(ctx *context.Context) {
+	// get task
+	taskID := ctx.ParamsInt64("taskID")
+	task, err := actions.GetTaskByID(ctx, taskID)
+	if err != nil {
+		log.Error("Error getting task: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	artifact, err := actions.CreateArtifact(ctx, task)
+	if err != nil {
+		log.Error("Error creating artifact: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+	url, err := ar.buildArtifactURL(taskID, artifact.ID, "upload")
+	if err != nil {
+		log.Error("Error parsing upload URL: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	log.Debug("[artifact] get upload url: %s, artifact id: %d", url, artifact.ID)
+
+	ctx.JSON(http.StatusOK, map[string]interface{}{
+		"fileContainerResourceUrl": url,
+	})
+}
+
+// getUploadFileSize returns the size of the file to be uploaded.
+// The raw size is the size of the file as reported by the header X-TFS-FileLength.
+func (ar artifactRoutes) getUploadFileSize(ctx *context.Context) (int64, int64, error) {
+	contentLength := ctx.Req.ContentLength
+	xTfsLength, _ := strconv.ParseInt(ctx.Req.Header.Get(artifactXTfsFileLengthHeader), 10, 64)
+	if xTfsLength > 0 {
+		return xTfsLength, contentLength, nil
+	}
+	return contentLength, contentLength, nil
+}
+
+func (ar artifactRoutes) saveUploadChunk(ctx *context.Context,
+	artifact *actions.ActionArtifact,
+	contentSize, taskID int64,
+) (int64, error) {
+	// read chunk data to calc md5
+	chunkData, err := io.ReadAll(ctx.Req.Body)
+	if err != nil {
+		return -1, err
+	}
+
+	// check md5
+	md5Hash := md5.Sum(chunkData)
+	md5String := base64.StdEncoding.EncodeToString(md5Hash[:])
+	if md5String != ctx.Req.Header.Get(artifactXActionsResultsMD5Header) {
+		return -1, fmt.Errorf("md5 not match")
+	}
+
+	contentRange := ctx.Req.Header.Get("Content-Range")
+	start, end, length := int64(0), int64(0), int64(0)
+	if _, err := fmt.Sscanf(contentRange, "bytes %d-%d/%d", &start, &end, &length); err != nil {
+		return -1, fmt.Errorf("parse content range error: %v", err)
+	}
+
+	storagePath := fmt.Sprintf("tmp%d/%d-%d-%d.chunk", taskID, artifact.ID, start, end)
+
+	// save chunk to storage
+	if _, err := ar.fs.Save(storagePath, bytes.NewBuffer(chunkData), -1); err != nil {
+		return -1, fmt.Errorf("save chunk to storage error: %v", err)
+	}
```

Comment:
To avoid read all bytes to memory:

```go
hasher := md5.New()
hr := &hashReader{
	Reader: mockReqBody,
	Hasher: hasher,
}
ar.fs.Save(hr)
md5sum := hr.Hasher.Sum(nil)
	
// TODO: if md5sum is mismatch, remove the stored object

type hashReader struct {
	Reader io.Reader
	Hasher hash.Hash
}

func (hr *hashReader) Read(p []byte) (n int, err error) {
	n, err = hr.Reader.Read(p)
	if err == nil {
		hr.Hasher.Write(p[:n])
	}
	return n, err
}
```